src/Raven.Server/Documents/ShardedTcpHandlers/ShardedSubscriptionConnection.cs

```diff
@@ -0,0 +1,583 @@
+                        ReadResponseAndGetVersionCallbackAsync = async (operationContext, writer, stream1, _) =>
+                        {
+                            using (var response = await operationContext.ReadForMemoryAsync(stream1, "Subscription/sharded/tcp-header-response").ConfigureAwait(false))
+                            {
+                                var reply = JsonDeserializationClient.TcpConnectionHeaderResponse(response);
+                                switch (reply.Status)
+                                {
+                                    case TcpConnectionStatus.Ok:
+                                        return reply.Version;
+                                    case TcpConnectionStatus.AuthorizationFailed:
+                                        throw new AuthorizationException($"Cannot access shard '{db}' because " + reply.Message);
+                                    case TcpConnectionStatus.TcpVersionMismatch:
+                                        if (reply.Version != TcpNegotiation.OutOfRangeStatus)
+                                            return reply.Version;
+
+                                        //Kindly request the server to drop the connection
+                                        operationContext.Write(writer, new DynamicJsonValue
+                                        {
+                                            [nameof(TcpConnectionHeaderMessage.DatabaseName)] = db,
+                                            [nameof(TcpConnectionHeaderMessage.Operation)] = TcpConnectionHeaderMessage.OperationTypes.Drop.ToString(),
+                                            [nameof(TcpConnectionHeaderMessage.OperationVersion)] =
+                                        TcpConnectionHeaderMessage.GetOperationTcpVersion(TcpConnectionHeaderMessage.OperationTypes.Drop),
+                                            [nameof(TcpConnectionHeaderMessage.Info)] = $"Couldn't agree on subscription TCP version ours: '{TcpConnectionHeaderMessage.SubscriptionTcpVersion}' theirs: '{reply.Version}'"
+                                        });
+
+                                        await writer.FlushAsync().ConfigureAwait(false);
+                                        throw new InvalidOperationException($"Can't connect to shard '{db}' because: {reply.Message}");
+                                }
+
+                                return reply.Version;
+                            }
+                        },
+                        DestinationUrl = chosenUrl
+                    });
+
+                    if (supportedFeatures.ProtocolVersion <= 0)
+                        throw new InvalidOperationException($"{Options.SubscriptionName}: TCP negotiation resulted with an invalid protocol version:{supportedFeatures.ProtocolVersion}");
+
+
+                    _workers.Add(new SubscriptionServerWorker(stream, tcpClient, _serverStore));
+                }
```

Comment:
That is a lot of duplicated code. We have that elsewhere, can we avoid this?

Reply:
this was moved to `AbstractSubscriptionWorker` which used by both client worker and shard worker